core/src/main/scala/com/databricks/labs/remorph/coverage/estimation/EstimationAnalyzer.scala

```diff
@@ -11,14 +14,159 @@ object SqlComplexity {
+    complexity: SqlComplexity)
+
+object EstimationStatistics {
+  implicit val rw: ReadWriter[EstimationStatistics] = macroRW
+}
+
 class EstimationAnalyzer extends LazyLogging {
   // TODO: We will do this and the rest of the analysis in the next PR
   def countStatements(plan: ir.LogicalPlan): Int = 1
+
+  def estimateComplexity(query: ExecutedQuery, plan: ir.LogicalPlan): ComplexityEstimate = {
+    val s = sourceTextComplexity(query.source)
+    ComplexityEstimate(SqlComplexity.LOW, countStatements(plan), s.textLength, s.lineCount)
+  }
+
+  def evaluateTree(node: TreeNode[_]): Int = {
+    evaluateTree(node, logicalPlanEvaluator, expressionEvaluator)
+  }
+
+  def evaluateTree(
+      node: TreeNode[_],
+      logicalPlanVisitor: PartialFunction[LogicalPlan, Int],
+      expressionVisitor: PartialFunction[Expression, Int]): Int = {
+    if (node == null) return 0 // Return 0 if the node is null
+
+    node match {
+      case lp: LogicalPlan =>
+        if (lp.children == null) return 0
+        val currentValue = if (logicalPlanVisitor.isDefinedAt(lp)) logicalPlanVisitor(lp) else 0
+        val childrenValue = lp.children.map(child => evaluateTree(child, logicalPlanVisitor, expressionVisitor)).sum
+        val expressionsValue = lp.expressions.map(expr => evaluateTree(expr, logicalPlanVisitor, expressionVisitor)).sum
+        currentValue + childrenValue + expressionsValue
+
+      case expr: Expression =>
+        if (expr.children == null) return 0
+        val currentValue = if (expressionVisitor.isDefinedAt(expr)) expressionVisitor(expr) else 0
+        val childrenValue = expr.children.map(child => evaluateTree(child, logicalPlanVisitor, expressionVisitor)).sum
+        currentValue + childrenValue
+
+      case _ =>
+        throw new IllegalArgumentException(s"Unsupported node type: ${node.getClass.getSimpleName}")
+    }
+  }
+
+  /**
+   * <p>
+   *   Given the raw query text, produce some statistics that are purely derived from the text, rather than
+   *   a parsed plan or translation result.
+   * </p>
+   * <p>
+   *   Text complexity is just one component for the overall score of a query, but it can be a good indicator
+   *   of how complex the query is in terms of a human translating it. For example, a query with many lines
+   *   and a lot of text is likely to take some time to manually translate, even if there are no complex
+   *   expressions, UDFs or subqueries. Text length is of little consequence to the transpiler if it is
+   *   successful in parsing but there is.
+   * </p>
+   *
+   * @param query the raw text of the query
+   * @return a set of statistics about the query text
+   */
+  def sourceTextComplexity(query: String): SourceTextComplexity = {
+    SourceTextComplexity(query.split("\n").length, query.length)
+  }
+
+  private def logicalPlanEvaluator: PartialFunction[LogicalPlan, Int] = { case lp: LogicalPlan =>
+    try {
+      1
+    } catch {
+      case _: Throwable => 5
+    }
+  }
```

Comment:
I don't see what exception could be thrown while evaluating `1`. Maybe this is just some placeholder code that will be filled later?

Also, catching `Throwable` is advised against. Instead, you should `catch NonFatal(e) =>`

Reply:
The 1 is indeed just a placeholder for the next PR. I see that some of the nodes come back as null (which likely means there are missing bits in IR building). I will change to catch NonFatal.